blocks/hooks/anchor.js

```diff
@@ -0,0 +1,77 @@
+/**
+ * External dependencies
+ */
+import { assign, get } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { cloneElement } from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import { source } from '../api';
+import InspectorControls from '../inspector-controls';
+
+/**
+ * Regular expression matching invalid anchor characters for replacement.
+ *
+ * @type {RegExp}
+ */
+const ANCHOR_REGEX = /[\s#]/g;
+
+export default function anchor( settings ) {
+	if ( ! get( settings.supports, 'anchor' ) ) {
+		return settings;
+	}
+
+	// Extend attributes with anchor determined by ID on the first node.
+	assign( settings.attributes, {
```

Comment:
Should we avoid to edit the function parameter, Should we make this pure instead?

Reply:
This sounds like something that would be difficult to enforce, and would have questionable value if not enforced consistently. If it _could_ be enforced, I could potentially see some value in strict reference comparisons on before/after a filter is applied (similar to what we see in how Redux reducer values are used).

As implemented here, we get some performance gains from mutating the object directly rather than first creating a shallow clone. I think it's reasonable to consider the filtering as an extension of the internal behavior of `registerBlockType`, in which we have no qualms to mutate the settings object directly.

I could go either way personally, but the immediate value isn't obvious to me.